nixos/modules/services/networking/rosenpass.nix

```diff
@@ -0,0 +1,178 @@
+    };
+
+    peers = let
+      peer = submodule {
+        options = {
+          publicKeyFile = mkOption {
+            type = path;
+            description = "Path to a file containing the public key of the remote Rosenpass peer.";
+          };
+
+          endpoint = mkOption {
+            type = nullOr str;
+            default = null;
+            description = "Endpoint of the remote Rosenpass peer.";
+          };
+
+          device = mkOption {
+            type = str;
+            default = cfg.defaultDevice;
+            defaultText = literalExpression "config.${opt.defaultDevice}";
+            description = "Name of the local WireGuard interface to use for this peer.";
+          };
+
+          wireguard = mkOption {
+            type = submodule {
+              options = {
+                publicKey = mkOption {
+                  type = str;
+                  description = "WireGuard public key corresponding to the remote Rosenpass peer.";
+                };
+              };
+            };
+            description = "WireGuard configuration for this peer.";
+          };
+        };
+      };
+    in
+      mkOption {
+        type = listOf peer;
+        description = "List of peers to exchange keys with.";
+        default = [];
+      };
+
+    extraConfig = mkOption {
+      type = attrs;
+      description = ''
+        Extra configuration to be merged with the generated Rosenpass configuration file.
+      '';
+      default = {};
+    };
+  };
+
+  config = mkIf cfg.enable {
+    warnings = let
+      netdevsList = attrValues config.systemd.network.netdevs;
+      publicKeyInNetdevs = peer: any (netdev: any (publicKeyInWireguardPeers peer) netdev.wireguardPeers) netdevsList;
+      publicKeyInWireguardPeers = peer: x: x.wireguardPeerConfig ? PublicKey && x.wireguardPeerConfig.PublicKey == peer.wireguard.publicKey;
+
+      # NOTE: In the message below, we tried to refer to
+      #   options.systemd.network.netdevs."<name>".wireguardPeers.*.PublicKey
+      # directly, but don't know how to traverse "<name>" and * in this path.
+      warningMsg = peer: "It appears that you have configured a Rosenpass peer with the Wireguard public key '${peer.wireguard.publicKey}' but there is no corresponding Wireguard peer configuration in any of `${options.systemd.network.netdevs}.\"<name>\".wireguardPeers.*.PublicKey`. While this may work as expected, such a scenario is unusual. Please double-check your configuration.";
+    in
+      concatMap (peer: optional (!publicKeyInNetdevs peer) (warningMsg peer)) cfg.peers;
+
+    environment.systemPackages = [cfg.package pkgs.wireguard-tools];
+
+    users.users."${cfg.user}" = {
+      isSystemUser = true;
+      createHome = false;
+      group = cfg.group;
+    };
+
+    users.groups."${cfg.group}" = {};
+
+    # NOTE: It would be possible to use systemd credentials for pqsk.
+    # <https://systemd.io/CREDENTIALS/>
```

Comment:
It would be a good idea to use `LoadCredential` by default. That will make it possible to use `DynamicUser` as commented elsewhere.